### PR TITLE
ceph: Add support for volume replication controller sidecar in Rook

### PR DESCRIPTION
**Pull request by bipuladh**
**Description of your changes:**
Component: Ceph
Adding a sidecar container to support Volume replication controller

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [x] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [x] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:123]: Why use an env var? If we always want to use the socket and it doesn't need to be configurable, it could be specified directly in the args.

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:127]: I don't see this volume defined.

File: cluster/charts/rook-ceph/values.yaml
Comment (travisn) [inline cluster/charts/rook-ceph/values.yaml:285]: What about this structure instead?
```yaml
volumeReplication:
  enabled: "false"
  image: "quay.io/kubestorage/volumereplication-operator:latest"
```

File: cluster/examples/kubernetes/ceph/operator.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator.yaml:278]: We should set this to true in some integration tests. Let me get back to you on the specifics after #7390 is merged.

**Review by travisn — CHANGES_REQUESTED** [review]

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (bipuladh) [reply cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:127]: https://github.com/rook/rook/pull/7403/files#diff-94612bfa354b036a2f57b1771f8b73f02d287d6a72e62b3dc21bcb3a0771d7edL184
isn't this the defn?

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (bipuladh) [reply cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:123]: following the pattern used in other places. Should I continue the pattern or directly put it in the args?

File: cluster/examples/kubernetes/ceph/operator.yaml
Comment (bipuladh) [reply cluster/examples/kubernetes/ceph/operator.yaml:278]: ack

File: cluster/examples/kubernetes/ceph/common.yaml
Comment (subhamkrai) [inline cluster/examples/kubernetes/ceph/common.yaml:1087]:
```suggestion
    verbs: ["create", "delete", "get", "list", "patch", "update", "watch"]
```

see yamllint is failing due to this.

**Review by subhamkrai — COMMENTED** [review]
and see [this](https://rook.io/docs/rook/master/development-flow.html#commit-structure) for commitlint error.

it should be something like
```
ceph:  add support for volume replication controller sidecar 

then the description of the changes
```
Thanks

**Comment by subhamkrai** [comment]
@bipuladh can you make sure ```

 header must not be longer than 72 characters, current length is 73 [header-max-length]
✖   subject must not be sentence-case, start-case, pascal-case, upper-case [subject-case]```
you can look at the error. Thanks

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (travisn) [reply cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:123]: Oh, same pattern sounds good

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (travisn) [reply cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:127]: I see it now, thanks

File: cluster/examples/kubernetes/ceph/operator.yaml
Comment (travisn) [reply cluster/examples/kubernetes/ceph/operator.yaml:278]: We can only enable it on openshift, right? The setting can be replaced in the operator.yaml, see the method `replaceOperatorSettings()` where you could call 
```go
if utils.IsPlatformOpenShift() {
  // replace the setting in the yaml
  manifest = strings.ReplaceAll(...)
}
```

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:116]: Is there a versioned tag instead of `latest`? We could get broken if there is an update to the tag for a change we don't support.

File: cluster/examples/kubernetes/ceph/common.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/common.yaml:1085]: These changes would also be needed in the helm chart in clusterrole.yaml.

File: cluster/examples/kubernetes/ceph/common.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/common.yaml:1088]: Would replication only work on openshift? Or is the RBAC just defined in this namespace at it will still work on K8s?

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:111]: @Madhu-1 We would also need these changes in the csi repo, right?

**Review by travisn — CHANGES_REQUESTED** [review]
Some documentation to consider:
- PendingReleaseNotes.md: Add a line with this feature
- ceph-csi-driver.md: Perhaps add a section about enabling this feature to help users understand if they want to enable it.

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:111]: Yes, this is needed in the cephcsi repo. Will work on cephcsi later

File: cluster/examples/kubernetes/ceph/common.yaml
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/common.yaml:1088]: This is just an API group it should work on kubernetes as well.

File: pkg/operator/ceph/csi/spec.go
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:116]: Yes, this PR should not get merged now, This is up for review. We can merge this one once we have replication operator release and cephcsi release. Adding DNM

**Review by Madhu-1 — COMMENTED** [review]
Requesting changes to avoid merging till we get new volume replication operator release and also this is dependent on coming cephcsi release

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
Requesting changes to avoid merging till we get new volume replication operator release and also this is dependent on coming cephcsi release

File: cluster/examples/kubernetes/ceph/operator.yaml
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/operator.yaml:278]: This can be enabled on kubernetes and OCP. this is not OCP specific

File: pkg/operator/ceph/csi/spec.go
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:68]: move this above where bool and strings are defined

File: pkg/operator/ceph/csi/spec.go
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:312]: `CSI_ENABLE_VOL_REPLICATION` can be changed to `CSI_ENABLE_VOLUME_REPLICATION`

File: pkg/operator/ceph/csi/csi.go
Comment (Madhu-1) [inline pkg/operator/ceph/csi/csi.go:112]: CSI_VOL_REPLICATION_IMAGE to CSI_VOLUME_REPLICATION_IMAGE

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:121]: this should be set to true and also add leader-election-namespace https://github.com/kube-storage/volume-replication-operator/blob/main/main.go#L68 which is pod namespace. check other sidecars for example

File: cluster/examples/kubernetes/ceph/common.yaml
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/common.yaml:1085]: ReplicationClass related RBAC is missing

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:359]: comment out these new fields in both operator.yaml and openshift-operator.yaml

File: design/ceph/ceph-csi-driver.md
Comment (Madhu-1) [inline design/ceph/ceph-csi-driver.md:149]: more details need to be added here on how to do failover and failback for DR. refer https://github.com/kube-storage/volume-replication-operator#design

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml
Comment (bipuladh) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:359]: https://github.com/rook/rook/pull/7403#discussion_r594561531 
This comment says otherwise.

**Comment by Madhu-1** [comment]
@travisn PTAL

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:361]: new image org `https://quay.io/repository/csiaddons/volumereplication-operator` need to be updated on other places also

File: design/ceph/ceph-csi-driver.md
Comment (Madhu-1) [inline design/ceph/ceph-csi-driver.md:149]:
```suggestion
Volume Replication Operator is a kubernetes operator that provides common and reusable APIs for storage disaster recovery. It is based on [csi-addons/spec](https://github.com/csi-addons/spec) specification and can be used by any storage provider.
```

File: pkg/operator/ceph/csi/spec.go
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:43]: VolumeReplicationImage is missing from this struct

File: PendingReleaseNotes.md
Comment (travisn) [inline PendingReleaseNotes.md:18]: Let's move this to the ceph section below

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:361]: Should the image be commented out by default and use a default from the code? This would be the same pattern as the CSI driver.

File: design/ceph/ceph-csi-driver.md
Comment (travisn) [inline design/ceph/ceph-csi-driver.md:148]:
```suggestion
To support RBD Mirroring, the [Volume Replication Operator](https://github.com/csi-addons/volume-replication-operator/blob/main/README.md) will be started in the RBD provisioner pod.
```

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:360]: We're disabling by default, right?
```suggestion
  CSI_ENABLE_VOLUME_REPLICATION: "false"
```

File: PendingReleaseNotes.md
Comment (travisn) [reply PendingReleaseNotes.md:18]: Does the replication controller require the CSI driver 3.3?

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:116]: Still concerned about the `latest` tag. We should use a version tag instead of a floating tag.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go
Comment (bipuladh) [reply pkg/operator/ceph/csi/spec.go:116]: There are no other tags presently.

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml
Comment (bipuladh) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:360]: Yes! But I was under the impression we should enable it for testing purposes.

Comment (travisn) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:360]: For testing purposes, instead of setting it in operator.yaml, we need to replace it in the test settings. See [replaceOperatorSettings()](https://github.com/rook/rook/blob/fcf634e3b7808e4338e7fd261c047aae4793e99b/tests/framework/installer/ceph_settings.go#L103-L109)

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [reply pkg/operator/ceph/csi/spec.go:116]: We own the image, right? Let's ask for versioned tags, even if v0.0.1 to get started.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:116]: `v0.1.0` image is available `quay.io/csiaddons/volumereplication-operator:v0.1.0`

File: PendingReleaseNotes.md
Comment (bipuladh) [reply PendingReleaseNotes.md:18]: Yes, should we mention that here as well?

Comment (Madhu-1) [reply PendingReleaseNotes.md:18]: Yes. For we can document it. Maybe I can send a patch later to do this validation.

File: cluster/charts/rook-ceph/values.yaml
Comment (Madhu-1) [inline cluster/charts/rook-ceph/values.yaml:285]: latest need to be changed to v0.1.0

this need to be documented at https://github.com/rook/rook/blob/master/Documentation/helm-operator.md#configuration

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:110]: This will conflict with csi 3.3.0 in #7637. How about we merge the other one first since this depends on it anyway?

File: design/ceph/ceph-csi-driver.md
Comment (travisn) [inline design/ceph/ceph-csi-driver.md:150]: Let's document how they would enable the replication controller, perhaps just to link to the helm doc and mention the env var in the operator.yaml if they aren't using helm

File: design/ceph/ceph-csi-driver.md
Comment (travisn) [inline design/ceph/ceph-csi-driver.md:146]:
```suggestion
### RBD Mirroring
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: PendingReleaseNotes.md
Comment (Madhu-1) [inline PendingReleaseNotes.md:38]: Cephcsi v3.3.0

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:361]: This will be CSI_VOLUME_REPLICATION_IMAGE ?

**Review by Madhu-1 — APPROVED** [review]
Small nit, lgtm

File: design/ceph/ceph-csi-driver.md
Comment (travisn) [inline design/ceph/ceph-csi-driver.md:153]: Actually let's remove this link to master
```suggestion
- For non-Helm deployments set `CSI_ENABLE_VOLUME_REPLICATION: "true"` in the operator.yaml
```

File: design/ceph/ceph-csi-driver.md
Comment (travisn) [inline design/ceph/ceph-csi-driver.md:152]:
```suggestion
- For Helm deployments see the [helm settings](helm-operator.md#configuration).
```

**Review by travisn — APPROVED** [review]

**Pull request merged**
